lgc/elfLinker/GlueShader.cpp

```diff
@@ -282,3 +325,82 @@ Function *FetchShader::createFetchFunc() {
+  }
+}
+
+// =====================================================================================================================
+// Get the string for this color export shader. This is some encoding or hash of the inputs to the
+// createColorExportShader function that the front-end client can use as a cache key to avoid compiling the same glue
+// shader more than once.
+StringRef ColorExportShader::getString() {
+  if (m_shaderString.empty()) {
+    m_shaderString =
+        StringRef(reinterpret_cast<const char *>(m_exports.data()), m_exports.size() * sizeof(ColorExportInfo)).str();
+  }
+  return m_shaderString;
+}
+
+// =====================================================================================================================
+// Get the symbol name of the main shader that this glue shader is prolog or epilog for
+StringRef ColorExportShader::getMainShaderName() {
+  return getEntryPointName(CallingConv::AMDGPU_PS, /*isFetchlessVs=*/true);
+}
+
+// =====================================================================================================================
+// Generate the IR module for the color export shader
+Module *ColorExportShader::generate() {
+  // Create the function.
+  Function *colorExportFunc = createColorExportFunc();
+
+  // Process each vertex input.
+  std::unique_ptr<FragColorExport> fragColorExport(new FragColorExport(&getContext()));
+  auto ret = cast<ReturnInst>(colorExportFunc->back().getTerminator());
+  BuilderBase builder(ret);
+
+  for (unsigned idx = 0; idx != m_exports.size(); ++idx) {
+    const ColorExportInfo &exp = m_exports[idx];
+    Argument *output = colorExportFunc->getArg(idx);
+    ExportFormat expFmt = m_exportFormat[exp.hwColorTarget];
+    fragColorExport->run(output, exp.hwColorTarget, ret, expFmt, exp.isSigned);
+  }
+  return colorExportFunc->getParent();
+}
+
+// =====================================================================================================================
+// Create module with function for the fetch shader. On return, the function contains only the code to copy the
+// wave dispatch SGPRs and VGPRs to the return value.
+Function *ColorExportShader::createColorExportFunc() {
+  // Create the module
+  Module *module = new Module("colorExportShader", getContext());
+  TargetMachine *targetMachine = m_lgcContext->getTargetMachine();
+  module->setTargetTriple(targetMachine->getTargetTriple().getTriple());
+  module->setDataLayout(targetMachine->createDataLayout());
+
+  // Get the function type. Its inputs are the outputs from the unlinked pixel shader are the inputs.
+  SmallVector<Type *, 16> entryTys;
+  for (const auto &exp : m_exports)
+    entryTys.push_back(exp.ty);
+  auto funcTy = FunctionType::get(Type::getVoidTy(getContext()), entryTys, false);
+
+  // Create the function. Mark SGPR inputs as "inreg".
+  StringRef funcName = getEntryPointName(CallingConv::AMDGPU_PS, /*isFetchlessVs=*/false);
```

Comment:
Per comments on #796, the PS should keep its normal name, and you can probably just make up any name you like here without anyone else needing to agree on what it is called.